mqtt/mqtt3/src/client/mod.rs

```diff
@@ -810,6 +818,16 @@ impl Error {
             _ => false,
         }
     }
+
+    fn is_connection_error(&self) -> bool {
+        match self {
+            Error::DecodePacket(crate::proto::DecodeError::Io(_))
```

Comment:
interesting... could you please add comments on the reasoning why these types of errors considered connection error? encode/decode looks like not related to networking

Reply:
it's based on some testing and from the code, for example if there is no connection and trying to send it will return IO error with WriteZero. I added Arnav to the PR if he can confirm this is correct. Also PingTimer I am not so sure should be connection erro.

Reply:
It's an annoying constraint of `tokio_util::codec` that the codec errors have to be `From<std::io::Error>` even though the codec code itself doesn't do any I/O, because tokio uses the same type to bubble up errors from writing the encoded bytes to the stream / reading the bytes to be decoded from the stream through the `Framed`.

I had it in the back of my mind to separate the codec from the network for this reason. Basically stop using `tokio_util::codec` and instead encode/decode manually. It means the client has to do the job of the `Framed` wrt handling partial reads and writes, but it at least allows it to differentiate between I/O and encoding errors.